// SPDX-FileCopyrightText: Â© 2025 Tenstorrent AI ULC
//
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

package tt.telemetry;

// The telemetry gRPC service
service TelemetryService {
  // Simple ping/pong to measure RTT
  rpc Ping(PingRequest) returns (PingResponse) {}

  // Query a metric by name
  rpc QueryMetric(QueryMetricRequest) returns (QueryMetricResponse) {}

  // Stream telemetry updates matching a query (server-side streaming)
  rpc StreamMetrics(QueryMetricRequest) returns (stream QueryMetricResponse) {}

  // List all known metric paths grouped by type
  rpc ListMetrics(ListMetricsRequest) returns (ListMetricsResponse) {}
}

// Ping request carrying a timestamp
message PingRequest {
  int64 timestamp = 1;  // Client timestamp (milliseconds since epoch)
}

// Pong response echoing the timestamp
message PingResponse {
  int64 timestamp = 1;  // Echo of the client's timestamp
}

// Query request for a specific metric
message QueryMetricRequest {
  string metric_query = 1;  // Query string for the metric (e.g., metric name)
}

// Individual metric result with path, timestamp, and value
message BoolMetricResult {
  string path = 1;         // Full metric path
  uint64 timestamp = 2;    // Timestamp when sampled (milliseconds since epoch)
  bool value = 3;          // Metric value
}

message UintMetricResult {
  string path = 1;         // Full metric path
  uint64 timestamp = 2;    // Timestamp when sampled (milliseconds since epoch)
  uint64 value = 3;        // Metric value
}

message DoubleMetricResult {
  string path = 1;         // Full metric path
  uint64 timestamp = 2;    // Timestamp when sampled (milliseconds since epoch)
  double value = 3;        // Metric value
}

message StringMetricResult {
  string path = 1;         // Full metric path
  uint64 timestamp = 2;    // Timestamp when sampled (milliseconds since epoch)
  string value = 3;        // Metric value
}

// Query response with arrays of typed results
message QueryMetricResponse {
  repeated BoolMetricResult bool_results = 1;
  repeated UintMetricResult uint_results = 2;
  repeated DoubleMetricResult double_results = 3;
  repeated StringMetricResult string_results = 4;
}

// List metrics request
message ListMetricsRequest {
  string metric_query = 1;  // Optional query string to filter metrics (empty = all metrics)
}

// List metrics response with metric paths grouped by type
message ListMetricsResponse {
  repeated string bool_metrics = 1;    // Paths of all bool metrics
  repeated string uint_metrics = 2;    // Paths of all uint metrics
  repeated string double_metrics = 3;  // Paths of all double metrics
  repeated string string_metrics = 4;  // Paths of all string metrics
}
